### ISSUE TITLE
v2.7.0 SNI change broke connections to plain IP addresses with SSLHandshakeException "Hostname or IP address is undefined."

### ISSUE THREAD

**Issue by hoxu** [http-kit/http-kit#560]
v2.7.0, more specifically commit e5c8caa0166320d7f0d8c8a384a5d53f6736b29d (PR #513), broke connections to IP addresses, causing an `SSLHandshakeException` with error `Hostname or IP address is undefined.`

Seems to happen when connecting to servers with or without SNI extension. I checked for SNI with:

```sh
IP=x.x.x.x; openssl s_client -servername $IP -tlsextdebug -connect $IP:443 2>&1 |grep 'server name'
TLS server extension "server name" (id=0), len=0
```

## Reproduction

Using an IP address for www.google.com as an example:

```sh
cat > repro.clj <<EOF
(require '[org.httpkit.client])

(println @(org.httpkit.client/get "https://216.58.209.196"))
EOF
```

Good:

```sh
$ clojure -Sdeps '{:deps {http-kit/http-kit {:mvn/version "2.6.0"}}}' -M repro.clj
```

Bad:

```sh
$ clojure -Sdeps '{:deps {http-kit/http-kit {:mvn/version "2.7.0"}}}' -M repro.clj 
{:opts {:method :get, :url https://216.58.209.196}, :error #error {
 :cause Hostname or IP address is undefined.
 :via
 [{:type javax.net.ssl.SSLHandshakeException
   :message Hostname or IP address is undefined.
   :at [sun.security.ssl.Alert createSSLException Alert.java 130]}
  {:type java.security.cert.CertificateException
   :message Hostname or IP address is undefined.
   :at [sun.security.util.HostnameChecker match HostnameChecker.java 97]}]
 :trace
 [[sun.security.util.HostnameChecker match HostnameChecker.java 97]
  [sun.security.ssl.X509TrustManagerImpl checkIdentity X509TrustManagerImpl.java 457]
  [sun.security.ssl.X509TrustManagerImpl checkIdentity X509TrustManagerImpl.java 431]
  [sun.security.ssl.AbstractTrustManagerWrapper checkAdditionalTrust SSLContextImpl.java 1501]
  [sun.security.ssl.AbstractTrustManagerWrapper checkServerTrusted SSLContextImpl.java 1445]
  [sun.security.ssl.CertificateMessage$T13CertificateConsumer checkServerCerts CertificateMessage.java 1296]
  [sun.security.ssl.CertificateMessage$T13CertificateConsumer onConsumeCertificate CertificateMessage.java 1195]
  [sun.security.ssl.CertificateMessage$T13CertificateConsumer consume CertificateMessage.java 1138]
  [sun.security.ssl.SSLHandshake consume SSLHandshake.java 393]
  [sun.security.ssl.HandshakeContext dispatch HandshakeContext.java 476]
  [sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction run SSLEngineImpl.java 1273]
  [sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction run SSLEngineImpl.java 1260]
  [java.security.AccessController doPrivileged AccessController.java 714]
  [sun.security.ssl.SSLEngineImpl$DelegatedTask run SSLEngineImpl.java 1205]
  [org.httpkit.client.HttpsRequest doHandshake HttpsRequest.java 91]
  [org.httpkit.client.HttpClient doRead HttpClient.java 220]
  [org.httpkit.client.HttpClient run HttpClient.java 521]
  [java.lang.Thread run Thread.java 1583]]}}
```

Still reproduces with:

```sh
clojure -Sdeps '{:deps {http-kit/http-kit {:mvn/version "2.8.0-beta3"}}}' -M repro.clj
```

**Comment by ptaoussanis** [comment]
@hoxu Thanks for the clear report Heikki, that's very helpful!

This is definitely an unintended breakage. PR welcome (from you or anyone else interested) that includes a fix and test/s for IP requests 👍

**Comment by ptaoussanis** [comment]
@hoxu I actually took a quick stab at this myself, could you please try `[http-kit "2.8.0-SNAPSHOT"]` (now on Clojars), and let me know if that resolves your issue?

Thanks!

**Comment by ptaoussanis** [comment]
> 2.8.0-SNAPSHOT seems to fix the project/code I'm working on, great!

Thanks for the confirmation 👍

I'm planning to release v2.8.0-RC1 before the end of this month, which will include the above fix.

> The repro code now also returns a different exception

Just to confirm: my understanding is that the new exception is the **correct/expected** behaviour. I.e. when connecting to an IP-based host over SSL, the certificate would need to specify that IP address.

Such certificates don't seem to be common, but there seems to be one from CloudFlare at https://1.1.1.1 which I've used in the tests.

Does that agree with your understanding? If so, please feel free to close!

Cheers :-)

**Comment by ptaoussanis** [comment]
> I would have expected the 2.7.0 case to throw the "Hostname or IP address is undefined." exception.

That was my expectation too - good catch, thanks. My best guess for why 2.7 isn't throwing here is that 1.1.1.1 and its certificate are configured to actually accept "1.1.1.1" as a valid SNI hostname. That's probably unusual, but indeed does seem to be possible.

I can't currently find an example of a publicly available certificate that behaves like your privately signed one. Ideas welcome if you know of any, otherwise it might be that if we want to test this properly, we'll have to use a custom certificate which'd be a bit more effort to set up.

PR welcome if you or anyone else feels like looking into that.

Otherwise I'm inclined to proceed with the planned release of 2.8.0-RC1 in the meantime, and someone can hopefully provide a working regression test in the future.

But do please let me know if you have another preference or idea. Cheers :-)

**Comment by ptaoussanis** [comment]
> But if it was, I guess one could test that the request throws SSLHandshakeException with a message "unable to find valid certification path to requested target".

That's a good idea, I've added such a test and released [v2.8.0-RC1](https://github.com/http-kit/http-kit/releases/tag/v2.8.0-RC1) 👍

Cheers!
